sdk/eventhub/Azure.Messaging.EventHubs/tests/Producer/EventHubProducerClientTests.cs

```diff
@@ -645,6 +646,525 @@ public async Task CloseAsyncDoesNotCloseTheConnectionWhenNotOwned()
+            var numberOfCalls = 0;
+
+            transportProducer
+                .Setup(transportProducer => transportProducer.SendAsync(It.IsAny<IEnumerable<EventData>>(),
+                                                                                It.IsAny<SendEventOptions>(),
+                                                                                It.IsAny<CancellationToken>()))
+                .Callback(() =>
+                {
+                    if (++numberOfCalls < EventHubProducerClient.MaximumCreateProducerAttempts)
+                    {
+                        throw new EventHubsException(false, string.Empty, EventHubsException.FailureReason.ClientClosed);
+                    }
+                })
+                .Returns(Task.CompletedTask);
+
+            Assert.That(async () => await producerClient.SendAsync(events, options), Throws.Nothing);
+        }
+
+        /// <summary>
+        ///   Verifies that the retry logic does not loop endlessly for batches.
+        /// </summary>
+        ///
+        [Test]
+        public void RetryLogicEndsWithABatch()
+        {
+            var batchOptions = new CreateBatchOptions { PartitionId = "0" };
+            var batch = new EventDataBatch(new MockTransportBatch(), batchOptions.ToSendOptions());
+            var transportProducer = new Mock<TransportProducer>();
+            var eventHubConnection = new MockConnection(() => transportProducer.Object);
+            var retryPolicy = new EventHubProducerClientOptions().RetryOptions.ToRetryPolicy();
+            var mockTransportProducerPool = new MockTransportProducerPool(new ObservableTransportProducerMock(), eventHubConnection, retryPolicy);
+            var mockPooledProducer = mockTransportProducerPool.GetPooledProducer(batchOptions.PartitionId) as MockPooledProducer;
+            var producerClient = new EventHubProducerClient(eventHubConnection, transportProducer.Object, mockTransportProducerPool);
+            var numberOfCalls = 0;
+
+            transportProducer
+                .Setup(transportProducer => transportProducer.SendAsync(It.IsAny<EventDataBatch>(),
+                                                                                It.IsAny<CancellationToken>()))
+                .Callback(() =>
+                {
+                    if (++numberOfCalls < EventHubProducerClient.MaximumCreateProducerAttempts)
+                    {
+                        throw new EventHubsException(false, string.Empty, EventHubsException.FailureReason.ClientClosed);
+                    }
+                })
+                .Returns(Task.CompletedTask);
+
+            Assert.That(async () => await producerClient.SendAsync(batch), Throws.Nothing);
+        }
+
+        /// <summary>
+        ///   Retry logic does not kick-in for the main-stream scenario.
+        /// </summary>
+        ///
+        [Test]
+        public void RetryLogicStartsForPartitionsOnly()
```

Comment:
nit: it took me some time to understand what's being tested here. The test name is comparing this behavior with another scenario instead of providing a proper description of what the test is about (the `summary` above was a big help). The same feedback applies to the batch version of the test.

I'm not entirely sure how the non-partition scenario should be called, though.

Reply:
Good point, what do you think of ```RetryLogicDoesNotStartWhenPartitionIdIsNull```?

Reply:
Sounds good to me!